File: src/main/java/com/lmp/catalog/domain/Service.java
package com.lmp.catalog.domain;

import com.lmp.billing.domain.CartItem;
import com.lmp.billing.domain.OrderItem;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "services")
public class Service {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private ServiceCategory category;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String description;

    private String icon;

    @Column(name = "display_order", nullable = false)
    private Integer displayOrder = 0;

    private Boolean featured = false;
    private Boolean active = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "service", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<ServiceBenefit> benefits;

    @OneToMany(mappedBy = "service", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<OrderItem> orderItems;

    @OneToMany(mappedBy = "service", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private Set<ServiceOffer> offers;

    @OneToMany(mappedBy = "service", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<CartItem> cartItems;

    public Service() {}

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }
    public ServiceCategory getCategory() { return category; }
    public void setCategory(ServiceCategory category) { this.category = category; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getIcon() { return icon; }
    public void setIcon(String icon) { this.icon = icon; }
    public String getSlug() { return slug; }
    public void setSlug(String slug) { this.slug = slug; }
    public Integer getDisplayOrder() { return displayOrder; }
    public void setDisplayOrder(Integer displayOrder) { this.displayOrder = displayOrder; }
    public Boolean getFeatured() { return featured; }
    public void setFeatured(Boolean featured) { this.featured = featured; }
    public Boolean getActive() { return active; }
    public void setActive(Boolean active) { this.active = active; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    public Set<ServiceBenefit> getBenefits() { return benefits; }
    public void setBenefits(Set<ServiceBenefit> benefits) { this.benefits = benefits; }
    public Set<OrderItem> getOrderItems() { return orderItems; }
    public void setOrderItems(Set<OrderItem> orderItems) { this.orderItems = orderItems; }
    public Set<ServiceOffer> getOffers() { return offers; }
    public void setOffers(Set<ServiceOffer> offers) { this.offers = offers; }
    public Set<CartItem> getCartItems() { return cartItems; }
    public void setCartItems(Set<CartItem> cartItems) { this.cartItems = cartItems; }

    @Transient
    public ServiceOffer getCurrentOffer() {
        if (offers == null || offers.isEmpty()) return null;
        ServiceOffer validPromo = offers.stream()
                .filter(ServiceOffer::isCurrentlyValid)
                .filter(o -> !Boolean.TRUE.equals(o.getIsDefault()))
                .findFirst().orElse(null);
        if (validPromo != null) return validPromo;
        return offers.stream()
                .filter(o -> Boolean.TRUE.equals(o.getActive()))
                .filter(o -> Boolean.TRUE.equals(o.getIsDefault()))
                .findFirst().orElse(null);
    }

    public void addBenefit(ServiceBenefit benefit) {
        if (benefits == null) benefits = new java.util.LinkedHashSet<>();
        benefits.add(benefit);
        benefit.setService(this);
    }

    // --- Champ de liaison système externe (agnostique ERP) ---

    @Column(name = "external_item_code", length = 140)
    private String externalItemCode;

    public String getExternalItemCode() { return externalItemCode; }
    public void setExternalItemCode(String externalItemCode) { this.externalItemCode = externalItemCode; }
}

File: backend-dotnet/src/Lmp.Domain/Catalog/Service.cs
namespace Lmp.Domain.Catalog;

/// <summary>
/// A catalogue service — maps to <c>services</c>. Aggregates its benefits and
/// priced offers, and resolves the "current" offer (valid promo &gt; default).
/// </summary>
public class Service
{
    public Guid Id { get; set; }
    public Guid CategoryId { get; set; }
    public ServiceCategory Category { get; set; } = null!;

    public string Title { get; set; } = null!;
    public string Slug { get; set; } = null!;
    public string Description { get; set; } = null!;
    public string? Icon { get; set; }
    public int DisplayOrder { get; set; }
    public bool Featured { get; set; }
    public bool Active { get; set; } = true;
    public DateTime? CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }

    /// <summary>Optional external-system link code (ERP-agnostic).</summary>
    public string? ExternalItemCode { get; set; }

    public ICollection<ServiceBenefit> Benefits { get; set; } = new List<ServiceBenefit>();
    public ICollection<ServiceOffer> Offers { get; set; } = new List<ServiceOffer>();

    /// <summary>
    /// The offer to display/charge: the first currently-valid non-default
    /// promotion, otherwise the active default offer. Mirrors
    /// <c>Service.getCurrentOffer()</c>.
    /// </summary>
    public ServiceOffer? GetCurrentOffer()
    {
        if (Offers.Count == 0)
        {
            return null;
        }

        var validPromo = Offers
            .Where(o => o.IsCurrentlyValid())
            .FirstOrDefault(o => !o.IsDefault);
        if (validPromo is not null)
        {
            return validPromo;
        }

        return Offers
            .Where(o => o.Active)
            .FirstOrDefault(o => o.IsDefault);
    }
}
